src/main/java/org/folio/bulkops/service/QueryService.java

```diff
@@ -1,94 +1,308 @@
+      operation.setErrorMessage(e.getMessage());
+      operation.setEndTime(LocalDateTime.now());
+      var linkToMatchingErrorsFile = errorService.uploadErrorsToStorage(operation.getId(), ERROR_MATCHING_FILE_NAME_PREFIX, e.getMessage());
+      operation.setLinkToMatchedRecordsErrorsCsvFile(linkToMatchingErrorsFile);
+    }
+  }
+
+  private void processQueryResult(QueryDetails queryDetails, String matchedCsvFileName, String matchedJsonFileName,
+                                  String matchedMrcFileName, BulkOperation operation) throws IOException, CsvRequiredFieldEmptyException, CsvDataTypeMismatchException {
+    try (var writerForResultCsvFile = remoteFileSystemClient.writer(matchedCsvFileName);
+         var writerForResultJsonFile = remoteFileSystemClient.writer(matchedJsonFileName);
+         var writerForResultMrcFile = remoteFileSystemClient.writer(matchedMrcFileName)) {
+      var entityClass = resolveEntityClass(operation.getEntityType());
+      var csvWriter = new BulkOperationsEntityCsvWriter(writerForResultCsvFile, entityClass);
+      List<BulkOperationExecutionContent> bulkOperationExecutionContents = new ArrayList<>();
+
+      int numMatched = 0;
+      int numProcessed = 0;
+      for (var content : queryDetails.getContent()) {
+
+        var entityRecord = constructRecord(operation, content);
+
+        try {
+          permissionsValidator.checkPermissions(operation, entityRecord);
+
+          var extendedRecord = constructExtendedRecord(operation, entityRecord);
+
+          if (entityRecord.isMarcInstance()) {
+            processQueryResultForMarc(entityRecord, writerForResultMrcFile, operation, matchedMrcFileName);
+          }
+
+          var extendedRecordAsJsonString = objectMapper.writeValueAsString(extendedRecord);
+
+          writerForResultJsonFile.append(extendedRecordAsJsonString);
+          CSVHelper.writeBeanToCsv(operation, csvWriter, extendedRecord.getRecordBulkOperationEntity(), bulkOperationExecutionContents);
+          numMatched++;
+        } catch (UploadFromQueryException e) {
+          handleError(bulkOperationExecutionContents, e, entityRecord, operation);
+        }
+        updateProgress(operation, numMatched, ++numProcessed);
+      }
+      errorService.saveErrorsAfterQuery(bulkOperationExecutionContents, operation);
+    }
+  }
+
+  private void updateProgress(BulkOperation operation, int numMatched, int numProcessed) {
+    operation.setProcessedNumOfRecords(numProcessed);
+    operation.setMatchedNumOfRecords(numMatched);
+    bulkOperationRepository.save(operation);
+  }
+
   private BulkOperation failBulkOperation(BulkOperation bulkOperation, String errorMessage) {
     bulkOperation.setStatus(FAILED);
     bulkOperation.setErrorMessage(errorMessage);
     bulkOperation.setEndTime(LocalDateTime.now());
-    var linkToCommittingErrorsFile = errorService.uploadErrorsToStorage(bulkOperation.getId(), ERROR_COMMITTING_FILE_NAME_PREFIX, bulkOperation.getErrorMessage());
```

Comment:
I think we need to fail job with saving of file with failure message. Please, consider saving operation failure but for matching stage.